README.md
# Virtual 6809 #

This project is a simple emulation of the Motorola 6809 8/16-bit processor in
javascript

In addition to the processor the project simulates the associated memory and 
interface chips that are required to interact with the "outside" world.

The Motorola reference design utilises a 64k memory model, a 6847 VDG, and 6821
PIA chips and while this project can simulate those chips it allows the use of 
other chip designs in the same role.

## Memory ##

The emulation of the memory is performed in blocks associated with specified 
memory address and ranges. For example the first 32k memory block can be 
defined as RAM, the next 16k as ROM, a gap of 8k followed by 8k of hardware 
mapped memory associated with the PIA.

In theory any memory model can be represented.
Memory bank swapping can also be achieved through use of a larger memory model
and appropriate hardware mapping.

For the purposes of emulation access to memory is independent of the clock
cycle allowing video hardware to read paged memory without interfering with the 
CPU interaction. This also means emulation of tri-stating the bus is not required and 
this in-turn means the emulated hardware is capable of some things that the real
CPU cannot do such as what would appear to be DMA actions from peripherals without
syncing the CPU.

## Processor ##

The CPU emulation is based on a strict read/process/write cpu cycle as detailed in the 
"6809 assembly language programming" book by Leventhal[1]. The aim is to achieve the same 
instruction timing.

The emulated processor can operate at any clock speed within the limits of the 
supporting hardware allowing for fine-tuning that was not possible with real 
hardware.

So far the instruction should now be complete. All instructions are tested but this is
still experimental so there may still be some wobbly bits

The ALU is fully implemented but condition logic contained in control 
register not ALU - this may be revised in future releases.  

The processor bootstraps correctly using the hard reset vector stored at 0xfffe.
If no memory exists at that address it will be interpreted as starting from 
0x0000.

## Video ##

The VDG is not an emulation, it is an approximation of the original hardware to
provide a more palatable experience than the 6847 reference hardware. The 
implication of this is that the multiplexing required to avoid bus contention is
not present.

VDG timing is independent of the actual cpu clock so display artifacts will be
radically different from the real components. The timing can be locked to the 
cpu clock if desired.

## Interaction ##

The project includes a budo'd client side application that simulates the cpu
running at approximately 1MHz

The hardware emulation does NOT include input yet so this is purely a passive
display, if the loaded program does not interact with video memory you will
not see anything happening

## How to Build and Run ##

The project is coded in Javascript using Node.js with budo. The unit tests are
composed using Jest

To "build" the project just copy the source or clone from github
Run NPM install from the project folder

The test suite can be started with `npm test`

To "run" the project use `npm start`

### Modifying the Hardware ###

The default memory hardware is an approximation of the coco/dragon design with a
32k lower RAM and the upper 32k a mix of ROM and mapped addresses (mapping goes
nowhere yet).

Changing `main.js` to alter the memory is simply a matter of changing these lines:

```javascript
const memory = Memory.factory("D64");
const machine = new Cpu(memory);
```

At the moment the memory factory only knows D64 and D4 but it is possible to assemble
a custom model using the memory manager:

```javascript
const memory = new MemoryManager([
  [new Chip(chips.RAM, chips.K32), 0x0000],
  [new Chip(chips.ROM, chips.K4), 0xf000]
]);
const machine = new Cpu(memory);
```
This would provide 4k of RAM at address 0 and a 4k ROM at 61440. Note it is advisable to 
provide a ROM segment at high-end of memory to provision the interrupt vector table.
Without some memory at this space the vector table will provide a 0 for all interrupts.

The composition of the memory (thanks to the vector table and mapped hardware) can be 
messy, but the memory mapping is on a first-come-first-served basis. If you specify an 
area of 256 bytes of mapped memory at 0xff00 and then 32kb of ram at 0x8000 the mapped
memory will take precedence at the high addresses. If you do this the other way around
it will be ram and not mapped.

The maximum expected memory model is 2Mb (the equivalent of using a 6829 MMU to provide
20 address lines) but paging is not implemented so in effect it will still only be 64k
of addressable memory.

### Trial Web App ###

The project is being published to v6809.herokuapp.com. This is not a persistent
app container so it may take a while to respond, please be patient!

### Sample Code ###
The "interactive" (?) demo of the emulator runs a simple piece of code to clear the
display, copy a message to the display and then infinitely cycle through one byte's
values on the screen:

```
const screenbase=$400

    org $2000
start:
    lds #$7fff
    ldx #screenbase
    bsr cls
    bsr message
    leax 32,x
cycle:
    inc ,x
    bra cycle
message:
    leay messagetext,pcr
    bsr print
    rts
cls:
    pshs a,x
    lda #$20
clsloop:
    sta ,x+
    cmpx #$0600
    bne clsloop
    puls a,x
    rts
print:
    pshs a,x,y
printloop:
    lda ,y+
    beq printover
print6847:
    sta ,x+
    bra printloop
printover:
    puls a,x,y
    rts

messagetext:
    fcb "VIRTUAL 6809 V0.4.3",0
```

The reset vector is set to the code origin so in theory it should always 
automatically run.

## References ##
[1] **6809 Assembly Language Programming**, 1981 by Lance A. Leventhal  
ISBN 0-931988-35-7

[2] **Motorola Semiconductor Technical Data - MC6809**

[3] **[Motorola MC6809-MC6809E 8-Bit Microprocessor Programming Manual](
https://www.maddes.net/m6809pm/sections.htm)**

[4] **[Paul Burgin's 6809 Instructions Crib Sheet](
https://techheap.packetizer.com/processors/6809/6809Instructions.html)**

CHANGELOG.md
# Virtual 6809 CHANGELOG #

## Unreleased ##

## V0.4.4 - Sept 17 2021 ##
* Update dependencies

## V0.4.3 - Feb 23 2021 ##
* Isolate production start script
* Create local run script (`npm run local`)
* Refactor code to play nicely with webpack

## V0.4.2 - Feb 23 2021 ##
* Change to default port number for Heroku

## V0.4.1 - Feb 23 2021 ##
* Fix byte order on pull pc
* Update sample code on demo

## V0.4.0 - Feb 22 2021 ##
* Add Indexed Direct Addressing
* Add Indexed Indirect Addressing
* Add CLR instructions

## V0.3.0 - Feb 18 2021 ##
__New:__
* Add EXG and TFR instructions
* Add Sign Extend (SEX) instruction
* Add MUL instruction
* Add TST instructions
* Add DAA instruction
* RTI instruction
* SWI instructions
* CWAI and SYNC instructions
* Interrupt handler

__Fixed:__
* Correction to stack alignment on pull

## V0.2.1 (fix) - Feb 13 2021 ##
* Use real semver to fix npm install

## V0.2 - Feb 11 2021 ##
* Add BIT, COM and NEG instructions
* Add shift and rotate instructions
* Add AND, OR, EOR instructions
* Add long branch instructions
* Add missing short branch instructions (BVS and BVC)
* A how to use section in the README
* Simplify and extend memory options

## V0.1.1 - Feb 03 2021 ##
* Fix incorrect operand code for JSR (was using BSR)
* Add BSR (using the right op code)

## V0.1.0 - Feb 03 2021 ##
* Minimal operand set
  * LOAD and SAVE
  * JMP
  * JSR, RTS, PSH and PUL
  * ADD and SUB (with and without carry)
  * INC and DEC
  * CMP (8 bit only)
  * ABX
  * NOP  
* Basic "interactive" client running at approximately 1MHz
  * Performance depends on host hardware but will not exceed 1MHz
* Flexible memory model 
  * 4k, 8k, 16k, 32k or 64k blocks

src/cpu/cpu_constants.js
const define = require('node-constants')(exports);
// register scale
define('SHORT', 8);
define('LONG', 16);
// cpu state
define('RUNNING', 1);
define('WAITING', 2);
define('SYNCING', 3);
// condition register flags
define('CARRY', 0x1);
define('OVERFLOW', 0x2);
define('ZERO', 0x4);
define('NEGATIVE', 0x8);
define('IRQ', 0x10);
define('HALFCARRY', 0x20);
define('FIRQ', 0x40);
define('ENTIRE', 0x80);
// interrupt lines
define('vSWI3', {
  'name': 'swi3',
  'vector': 0xfff2,
  'entire': true,
  'flags': 0x00,
});
define('vSWI2', {
  'name': 'swi2',
  'vector': 0xfff4,
  'entire': true,
  'flags': 0x00,
});
define('vFIRQ', {
  'name': 'firq',
  'vector': 0xfff6,
  'entire': false,
  'flags': this.IRQ | this.FIRQ,
});
define('vIRQ', {
  'name': 'irq',
  'vector': 0xfff8,
  'entire': true,
  'flags': this.IRQ,
});
define('vSWI', {
  'name': 'swi',
  'vector': 0xfffa,
  'entire': true,
  'flags': this.IRQ | this.FIRQ,
});
define('vNMI', {
  'name': 'nmi',
  'vector': 0xfffc,
  'entire': true,
  'flags': this.IRQ | this.FIRQ,
});
define('vRESET', {
  'name': 'reset',
  'vector': 0xfffe,
  'entire': true,
  'flags': this.IRQ | this.FIRQ,
});
// cpu pseudo-micro instructions
define('NEXT', 0); // get new instruction
define('FETCH', 1); // get next byte of instruction
define('DIRECT', 2); // generate direct page address
define('TFRWTOOB', 3); // transfer W to object
define('TFRWTOTG', 4); // transfer W to target
define('WRITEHIGH', 5); // write data high byte
define('WRITELOW', 6); // write data low byte
define('READHIGH', 7); // read high byte to object
define('READLOW', 8); // read low byte to object
define('READADHIGH', 9); // read high byte to AD
define('READADLOW', 10); // read low byte to AD
define('ADDTGTOOB', 11); // add b to object, result in W
define('BUSY', 12); // busy doing nothing
define('READWLOW', 13); // read low byte to W
define('READLOWCOMPARE', 14); // read low byte to W and compare with object
define('READADLOWCOMPARE', 15); // read byte from AD to w, compare with object
define('COMPAREW', 16); // compare W against object
define('READADWLOW', 17); // read low byte to W from AD
define('INCOB', 18); // increment object and place result in w
define('DECOB', 19);
define('INCW', 20);
define('DECW', 21);
define('WRITEWLOW', 22);
define('TFROBTOTG', 23);
define('TFRTGTOOB', 24);
define('PUSH', 25);
define('PULL', 26);
define('CUEPC', 27);
define('ADDPCTOOB', 28); // add byte at pc address to object + increment pc
define('ADDTGBTOOB', 29); // add byte/word at target address to object
define('SUBPCFROMOB', 30); // sub byte at pc address from object + increment pc
define('SUBTGFROMOB', 31); // sub byte/word at target address from object
define('SWAPWAD', 32); // swap internal registers
define('PUSHPC', 33);
define('PULLPC', 34);
define('SUBCPCFROMOB', 35); // sub byte at pc address from object with carry + increment pc
define('SUBCTGFROMOB', 36); // sub byte/word at target address from object with carry
define('ADDCPCTOOB', 37); // add byte at pc address to object with carry + increment pc
define('ADDCTGTOOB', 38); // add byte/word at target address to object with carry
define('ADDTGSTOOBIF', 39); // add target signed byte value to object if condition met
define('ADDTGSWTOOBIF', 40); // add target signed byte value to object if condition met
define('READAND', 41);
define('ANDCC', 42);
define('READOR', 43);
define('ORCC', 44);
define('READEOR', 45);
define('SHIFTLEFT', 46);
define('SHIFTRIGHT', 47);
define('ROTATELEFT', 48);
define('ROTATERIGHT', 49);
define('BITTEST', 50);
define('COMPLEMENT', 51);
define('NEGATE', 52);
define('EXCHANGE', 53);
define('TRANSFER', 54);
define('SIGNEXTEND', 55);
define('MULTIPLY', 56);
define('TESTOB', 57);
define('DECIMALADJUST', 58);
define('PULLCC', 59);
define('SETENTIRE', 60);
define('VECTORHIGH', 61);
define('VECTORLOW', 62);
define('MASKIF', 63);
define('WAIT', 64);
define('PUSHIR', 65);
define('SYNC', 66);
define('INDEX', 67);
define('CLEAR', 68);
define('SUBWFROMOB', 69);